docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,424 @@
+  Reason string `json:"reason"`
+}
+
+// LastError stores details of the most recent error encountered for the task.
+type LastError struct {
+  // Code is an error code that uniquely identifies an error.
+  Code ErrorCode `json:"code"`
+  // Description is a human-readable message indicating details of the error.
+  Description string `json:"description"`
+  // ObservedAt is the time at which the error was observed.
+  ObservedAt metav1.Time `json:"observedAt"`
+}
+
+// TaskState represents the state of the task.
+type TaskState string
+
+const (
+  Failed TaskState = "Failed"
+  Pending TaskState = "Pending"
+  Rejected TaskState = "Rejected"
+  Succeeded TaskState = "Succeeded"
+  InProgress TaskState = "InProgress"
+)
+
+// OperationState represents the state of last operation.
+type OperationState string
+
+const (
+  Failed OperationState = "OperationStateFailed"
+  Pending OperationState = "OperationStatePending"
+  Completed OperationState = "OperationStateCompleted"
+  InProgress OperationState = "OperationStateInProgress"
+)
+```
+
+### Custom Resource API object
+
+```yaml
+apiVersion: druid.gardener.cloud/v1alpha1
+kind: EtcdOperatorTask
+metadata:
+    name: <name of operator task resource>
+    namespace: <cluster namespace>
+    generation: <specific generation of the desired state>
+spec:
+    type: <type/category of supported out-of-band task>
+    ttlSecondsAfterFinished: <time-to-live to garbage collect the custom resource after it has been completed>
+    config: <task specific configuration>
+    ownerEtcdRefrence: <refer to corresponding etcd owner name and namespace for which task has been invoked>
+status:
+    observedGeneration: <specific observedGeneration of the resource>
+    state: <last known current state of the out-of-band task>
+    initiatedAt: <time at which task move to any other state from "pending" state>
+    lastErrors:
+    - code: <error-code>
+      description: <description of the error>
+      observedAt: <time the error was observed>
+    lastOperation:
+      name: <operation-name>
+      state: <task state as seen at the completion of last operation>
+      lastTransitionTime: <time of transition to this state>
+      reason: <reason/message if any>
+```
+
+### Lifecycle
+
+#### Creation
+
+Task(s) can be created by creating an instance of the `EtcdOperatorTask` custom resource specific to a task.
+
+> Note: In future, either a `kubectl` extension plugin or a `druidctl` tool will be introduced. Dedicated sub-commands will be created for each `out-of-band` task. This will drastically increase the usability for an operator for performing such tasks, as the CLI extension will automatically create relevant instance(s) of `EtcdOperatorTask` with the provided configuration.
+
+#### Execution
+
+* Authors propose to introduce a new controller (let's call it `operator-task-controller`) which watches for `EtcdOperatorTask` custom resource.
```

Comment:
```suggestion
* Authors propose to introduce a new controller which watches for `EtcdOperatorTask` custom resource.
```

Reply:
Since there is just one controller, one need not name it. The suggested name `operator-task-controller` is also not the one that we will eventually use. So i would refrain from naming it. If we have more than one controller then this would be fine.